### PR TITLE
GS Volume Culling

### PR DESCRIPTION
**Pull request by CedricGuillemet**
This is the 1st step for performance improvements with GS.
AABB Volumes used are coming from Compressed .PLY quantized chunks. It's far for being efficient. But goal here is to have everything in place to try different algorithms while not increasing complexity too much at once.

Batch of 256 splats are culled. visible ones are sorted and that subset is used for instance count.
It should decrease the pressure on GPU/CPU a little.

Next:
- Better partitioning (octree)
- wasm sorting
- radix sort?
- per splat culling
- per volume sorting of splats and volume sorting

EDIT: there should be 1 noticeable issue when camera moves fast with huge number of splats. Latency induced by sorting will make visible splat being not culled late and popping might be visible.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15705/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15705/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15705/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15705/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15705/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15705/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15705/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15705/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15705/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts [hunk 1]
```diff
@@ -13,13 +13,21 @@ import { Constants } from "core/Engines/constants";
...
+import { Frustum } from "core/Maths/math.frustum";
 
 interface DelayedTextureUpdate {
     covA: Float32Array;
     covB: Float32Array;
     colors: Float32Array;
     centers: Float32Array;
 }
+
+export interface BoundingVolume {
```
Comment (Popov72) [inline packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts:25]: I think `BoundingVolume` is a bit too generic a name, maybe prefix it with `GaussianSplatting` or something similar?

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts [hunk 2]
```diff
@@ -602,13 +697,18 @@ export class GaussianSplattingMesh extends Mesh {
...
+        this._worker.postMessage({ positions, vertexCount, bounds: boundingVolumesArrays.bounds, indices: boundingVolumesArrays.indices }, [
+            positions.buffer,
+            boundingVolumesArrays.bounds.buffer,
+            boundingVolumesArrays.indices.buffer,
+        ]);
 
         this._worker.onmessage = (e) => {
             this._depthMix = e.data.depthMix;
             const indexMix = new Uint32Array(e.data.depthMix.buffer);
+            //Logger.Log(`ahahahah ${this._vertexCount} - ${e.data.subsetSize}`);
```
Comment (Popov72) [inline packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts:709]:
```suggestion
```

File: packages/dev/loaders/src/SPLAT/splatFileLoader.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import type { ISceneLoaderPluginAsync, ISceneLoaderPluginFactory, ISceneLoaderAsyncResult, ISceneLoaderProgressEvent, SceneLoaderPluginOptions } from "core/Loading/sceneLoader";
 import { registerSceneLoaderPlugin } from "core/Loading/sceneLoader";
 import { SPLATFileLoaderMetadata } from "./splatFileLoader.metadata";
-import { GaussianSplattingMesh } from "core/Meshes/GaussianSplatting/gaussianSplattingMesh";
+import { GaussianSplattingMesh, type BoundingVolume } from "core/Meshes/GaussianSplatting/gaussianSplattingMesh";
```
Comment (Popov72) [inline packages/dev/loaders/src/SPLAT/splatFileLoader.ts:4]:
```suggestion
import type { BoundingVolume } from "core/Meshes/GaussianSplatting/gaussianSplattingMesh";
import { GaussianSplattingMesh } from "core/Meshes/GaussianSplatting/gaussianSplattingMesh";
```

**Review by Popov72 — CHANGES_REQUESTED** [review]

**Comment by CedricGuillemet** [comment]
Switching back to draft, I'm not sure CPU culling on a webworker is a good way to improve perf for high splat count.

**Pull request closed without merging**